catalogd/internal/storage/localdir.go

```diff
@@ -2,113 +2,316 @@ package storage
+	if err != nil {
+		return false
+	}
+	if !catalogFileStat.Mode().IsRegular() {
+		// path is not valid content
+		return false
+	}
+
+	if s.EnableQueryHandler {
+		indexFileStat, err := os.Stat(catalogIndexFilePath(s.catalogDir(catalog)))
+		if err != nil {
+			return false
+		}
+		if !indexFileStat.Mode().IsRegular() {
+			return false
+		}
+	}
+	return true
+}
+
+func (s *LocalDirV1) catalogDir(catalog string) string {
+	return filepath.Join(s.RootDir, catalog)
+}
+
+func catalogFilePath(catalogDir string) string {
+	return filepath.Join(catalogDir, "catalog.jsonl")
+}
+
+func catalogIndexFilePath(catalogDir string) string {
+	return filepath.Join(catalogDir, "index.json")
+}
+
+type storeMetasFunc func(catalogDir string, metaChan <-chan *declcfg.Meta) error
+
+func storeCatalogData(catalogDir string, metas <-chan *declcfg.Meta) error {
+	f, err := os.Create(catalogFilePath(catalogDir))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	for m := range metas {
+		if _, err := f.Write(m.Blob); err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
-func (s LocalDirV1) Delete(catalog string) error {
-	return os.RemoveAll(filepath.Join(s.RootDir, catalog))
+func storeIndexData(catalogDir string, metas <-chan *declcfg.Meta) error {
+	idx := newIndex(metas)
+
+	f, err := os.Create(catalogIndexFilePath(catalogDir))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	enc := json.NewEncoder(f)
+	enc.SetEscapeHTML(false)
+	return enc.Encode(idx)
 }
 
-func (s LocalDirV1) BaseURL(catalog string) string {
+func (s *LocalDirV1) BaseURL(catalog string) string {
 	return s.RootURL.JoinPath(catalog).String()
 }
 
-func (s LocalDirV1) StorageServerHandler() http.Handler {
+func (s *LocalDirV1) StorageServerHandler() http.Handler {
 	mux := http.NewServeMux()
-	fsHandler := http.FileServer(http.FS(&filesOnlyFilesystem{os.DirFS(s.RootDir)}))
-	spHandler := http.StripPrefix(s.RootURL.Path, fsHandler)
-	gzHandler := gzhttp.GzipHandler(spHandler)
 
-	typeHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+	mux.HandleFunc(s.RootURL.JoinPath("{catalog}", "api", "v1", "all").Path, s.handleV1All)
+	if s.EnableQueryHandler {
+		mux.HandleFunc(s.RootURL.JoinPath("{catalog}", "api", "v1", "query").Path, s.handleV1Query)
+	}
+	allowedMethodsHandler := func(next http.Handler, allowedMethods ...string) http.Handler {
+		allowedMethodSet := sets.New[string](allowedMethods...)
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if !allowedMethodSet.Has(r.Method) {
+				http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
+				return
+			}
+			next.ServeHTTP(w, r)
+		})
+	}
+	return allowedMethodsHandler(mux, http.MethodGet, http.MethodHead)
+}
+
+func (s *LocalDirV1) handleV1All(w http.ResponseWriter, r *http.Request) {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
+	catalog := r.PathValue("catalog")
+	catalogFile, catalogStat, err := s.catalogData(catalog)
```

Comment:
Need to `defer catalogFile.Close()` here.

Even better, now that we've got separate handling for all and query, maybe we go all the way to simply `http.ServeFile`?